Issue/11508 coupons card 2

hichamboushaba opened the pull request:
<!-- Remember about a good descriptive title. -->

Part of: #11508 
<!-- Id number of the GitHub issue this PR addresses. -->

### Description
This PR adds the following to the coupons Card:
- Loading UI.
- Empty View.
- Coupons List UI.
- Navigation to the coupons list.
- Navigation to the coupon details when a coupon is clicked.

### Testing instructions
1. Prepare some test data: create some coupons, and add them to some orders (you can do it from wp-admin even for old orders).
2. Open the app.
3. Switch to different date ranges, and confirm the coupons are loaded, and that the empty view is shown when needed.
4. Click on "View all coupons"
5. Confirm the coupons list is shown.
6. Go back, and click on one of the coupons.
7. Confirm the coupon details screen is shown.


### Images/gif
[Screen_recording_20240519_123512.webm](https://github.com/woocommerce/woocommerce-android/assets/1657201/12fae0bc-d4ae-4791-ba2c-133cb4e99f68)

- [x] I have considered if this change warrants release notes and have added them to `RELEASE-NOTES.txt` if necessary. Use the "[Internal]" label for non-user-facing changes.

<!-- Pull request guidelines: https://github.com/woocommerce/woocommerce-android/blob/develop/docs/pull-request-guidelines.md -->

hichamboushaba committed:
Update UI model with needed properties

hichamboushaba committed:
Add a loading UI

hichamboushaba committed:
Add UI for the coupon items

hichamboushaba committed:
Handle navigation to coupons and add menu

hichamboushaba committed:
Error handling

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

hichamboushaba committed:
Handle navigation to coupon details

hichamboushaba committed:
Update paddings

hichamboushaba committed:
Fix tint of the more icon in the coupon details

hichamboushaba committed:
Add an empty view to the coupons card

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/coupons/details/CouponDetailsScreen.kt:
```diff
@@ -83,9 +83,9 @@ fun CouponDetailsScreen(
+                        imageVector = Icons.Filled.MoreVert,
                         contentDescription = "Coupons Menu",
-                        tint = colorResource(id = R.color.action_menu_fg_selector)
+                        tint = MaterialTheme.colors.primary
```
Just an inconsistency that I noticed, we were using the wrong color for the menu button here.

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/coupons/DashboardCouponsCard.kt:
```diff
@@ -75,6 +100,33 @@ private fun HandleEvents(
+                }
+
+                is DashboardCouponsViewModel.ViewCouponDetails -> {
+                    navController.navigateSafely(
+                        DashboardFragmentDirections.actionDashboardToCouponListFragment()
+                    )
+                    navController.navigateSafely(
+                        directions = CouponListFragmentDirections.actionCouponListFragmentToCouponDetailsFragment(
+                            couponId = event.couponId
+                        ),
+                        skipThrottling = true,
+                        navOptions = navOptions {
+                            popUpTo(R.id.dashboard)
+                            anim {
+                                enter = R.anim.default_enter_anim
+                                exit = R.anim.default_exit_anim
+                                popEnter = R.anim.default_pop_enter_anim
+                                popExit = R.anim.default_pop_exit_anim
+                            }
+                        }
```
Instead of duplicating the couponDetailsFragment destination, I decided to just perform two navigation actions to show the details.

0nko replied:
I noticed that when I tap on a coupon and go back, the list isn't shown like for the reviews, but I'm back at the dashboard screen.

hichamboushaba replied:
Yes, that's intentional, since we don't have to handle the `undo` behavior like in the reviews.

hichamboushaba committed:
Fix detekt issues

hichamboushaba commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/coupons/DashboardCouponsCard.kt:
```diff
@@ -92,38 +144,62 @@ private fun DashboardCouponsCard(
+                is State.Error -> {
                     WidgetError(
-                        onContactSupportClicked = { /*TODO*/ },
-                        onRetryClicked = { /*TODO*/ }
+                        onContactSupportClicked = onContactSupportClick,
+                        onRetryClicked = onRetryClick
                     )
                 }
```
Here, we will need to differentiate between the two types of errors: a generic error, or when WCAdmin is disabled, this differentiation will be added during the work on https://github.com/woocommerce/woocommerce-android/issues/11544

hichamboushaba committed:
Merge branch 'trunk' into issue/11508-coupons-card-2

0nko reviewed (commented):
I noticed a weird behavior when I deleted a coupon:

1. I tapped on one of the coupons
2. Deleted it from the coupon details menu
3. When the app navigates back to the dashboard, the card shows the error screen and it's not possible to reload it

hichamboushaba committed:
Improve logic to account for coupon deletion

This updates the logic to load 6 coupons instead of 3, then exclude deleted coupons, and map only the rest.

hichamboushaba commented:
Thanks @0nko for the review, and thanks for catching the mentioned issue, it should be fixed now, please check p1716911641960219/1716911637.747699-slack-C03L1NF1EA3 for an internal discussion around the solution.

0nko reviewed (approved):
Thanks for addressing the issue, nice job! :shipit:
